forte/data/data_store.py

```diff
@@ -493,6 +496,137 @@ def get_entry(self, tid: int) -> Tuple[List, int, int]:
         # Otherwise, use ``index_id`` to find the index of the entry.
         raise NotImplementedError
 
+    def co_iterator_annotation_like(
+        self, type_names: List[str]
+    ) -> Iterator[List]:
+        r"""
+        Given two or more type names, iterate their entry lists from beginning to end together.
+
+        For every single type, their entry lists are sorted by the ``begin`` and
+        ``end`` fields. The ``co_iterator_annotation_like`` function will iterate those sorted lists
+        together, and yield each entry in sorted order. This tasks is quite
+        similar to merging several sorted list to one sorted list. We internally
+        use a `MinHeap` to order the order of yielded items, and the ordering
+        is determined by:
+
+            - start index of the entry.
+            - end index of the entry.
+            - the index of the entry type name in input parameter ``type_names``.
+
+        The precedence of those values indicates their priority in the min heap
+        ordering.
+        For example, if two entries have both the same begin and end field,
+        then their order is
+        decided by the order of user input type_name (the type that first
+        appears in the target type list will return first).
+        For entries that have the exact same `begin`, `end` and `type_name`,
+        the order will be determined arbitrarily.
+
+        Args:
+            type_names (List[str]): a list of string type names
+
+        Returns:
+
+            An iterator of entry elements.
+
+        """
+
+        n = len(type_names)
+        # suppose the length of type_names is N and the length of entry list of
+        # one type is M
+        # then the time complexity of using min-heap to iterate
+        # is O(M*log(N))
+
+        # Initialize the first entry of all entry lists
+        # it avoids empty entry lists or non-existant entry list
+        first_entries = []
+        for tn in type_names:
+            if tn not in self.__elements:
+                raise ValueError(
+                    f"Input parameter types name {tn} is not"
+                    "available. Please input available ones in this DataStore"
+                    f"object: {list(self.__elements.keys())}"
+                )
+
+        for tn in type_names:
+            # raise error if it's empty list
+            if not self.__elements[tn]:
+                raise IndexError(
+                    f"Entry list of type name ({tn}) is"
+                    " empty. Please check data in this DataStore "
+                    " to see if empty list is expected"
+                    f" or remove {tn} from input parameter type_names"
+                )
+            else:
+                first_entries.append(self.__elements[tn][0])
```

Comment:
What I am suggesting is actually 
```
        for tn in type_names:
           element_list = []
           try:
               element_list = self.__elements[tn]               
               first_entries.append(element_list[0])               
           except KeyError: 
               raise ValueError(
                    f"Input argument `type_names` to the function contains"
                    f" a type name [{tn}], which is not recognized."
                    f" Please input available ones in this DataStore"
                    f" object: {list(self.__elements.keys())}"
                )
          except IndexError:
                raise IndexError(
                    f"Entry list of type name ({tn}) is"
                    " empty. Please check data in this DataStore "
                    " to see if empty list is expected"
                    f" or remove {tn} from input parameter type_names"
                )
```

Reply:
@qinzzz @wanglec could you comment on whether `self.__elements[tn]` can be an empty list. Consider a scenario where you delete all elements of that type, will this list entry be deleted? Do we need to revisit the `delete` logic?

I would advise that `self.__elements[tn]` won't be an empty list, and the key should be deleted. This not also help making the serialization case (no point to serialize information for empty lists).

Reply:
I think it makes sense to delete the empty list and the corresponding key in `delete_entry`.

Reply:
@feipenghe let's make modifications in `delete_entry` to ensure that.

Reply:
Edited it.